### PR TITLE
JSONB Related improvements for PG 15

### PR DESCRIPTION
**Pull request by onderkalaci**
PG 15 introduced several ([[1]](https://www.depesz.com/2022/04/08/waiting-for-postgresql-15-documentation-for-sql-json-features/), [[2]](https://www.depesz.com/2022/04/06/waiting-for-postgresql-15-json_table/), [[3]](https://www.depesz.com/2022/04/01/waiting-for-postgresql-15-sql-json-query-functions/), [[4]](https://www.depesz.com/2022/04/01/waiting-for-postgresql-15-is-json-predicate/), [[5]](https://www.depesz.com/2022/03/31/waiting-for-postgresql-15-sql-json-constructors/)) features.  This PR aims to add support for those against `pg15_support` branch.

Majority of the changes just work. The main code change is around multi-shard support for `JSON_TABLE`. Now that we treat it as if it is a correlated function call. Luckily the `recurring tuples` infrastructure covers that nicely.

### CODE REVIEW

File: src/test/regress/expected/json_table_arbitrary_config.out [hunk 1]
```diff
@@ -0,0 +1,1562 @@
+SET search_path TO "json table";
```
Comment (onderkalaci) [inline src/test/regress/expected/json_table_arbitrary_config.out:1]: needs version check

File: src/backend/distributed/planner/multi_logical_planner.c [hunk 2]
```diff
@@ -1154,7 +1154,8 @@ HasComplexRangeTableType(Query *queryTree)
 		if (rangeTableEntry->rtekind != RTE_RELATION &&
 			rangeTableEntry->rtekind != RTE_SUBQUERY &&
 			rangeTableEntry->rtekind != RTE_FUNCTION &&
-			rangeTableEntry->rtekind != RTE_VALUES)
+			rangeTableEntry->rtekind != RTE_VALUES &&
+			!IsJsonTableRTE(rangeTableEntry))
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_logical_planner.c:1158]: We can probably handle arbitrary RTE_TABLEFUNC, but I'm ok with the scope.

File: src/backend/distributed/planner/query_pushdown_planning.c [hunk 3]
```diff
@@ -1805,17 +1804,20 @@ ContainsRecurringRangeTable(List *rangeTable, RecurringTuplesType *recurType)
 							  QTW_EXAMINE_RTES_BEFORE);
 }
 
+
 bool
```
Comment (marcocitus) [inline src/backend/distributed/planner/query_pushdown_planning.c:1808]: maybe comment, e.g.

IsJsonTableRTE checks whether the RTE refers to a JSON_TABLE table function,
which was introduced in PostgreSQL 15.

**Review by marcocitus — APPROVED** [review]

File: src/test/regress/expected/json_table_arbitrary_config.out (hunk 1 above)
Comment (onderkalaci) [reply src/test/regress/expected/json_table_arbitrary_config.out:1]: Also, maybe add 1-2 modification queries involving JSON_TABLE for sanity check

**Pull request merged**